numpyro/infer/mcmc.py

```diff
@@ -615,6 +621,257 @@ def __init__(self,
+        inv_mass_matrix_sqrt = np.linalg.cholesky(inverse_mass_matrix) if dense_mass \
+            else np.sqrt(inverse_mass_matrix)
+        if adapt_state_size is None:
+            # XXX: heuristic choice
+            adapt_state_size = 2 * z_flat.shape[-1]
+        else:
+            assert adapt_state_size > 1, 'adapt_state_size should be greater than 1.'
+        # NB: mean is init_params
+        zs = z_flat + _sample_proposal(inv_mass_matrix_sqrt, rng_key_zs, (adapt_state_size,))
+        # compute potential energies
+        pes = lax.map(lambda z: pe_fn(unravel_fn(z)), zs)
+        if dense_mass:
+            cov = np.cov(zs, rowvar=False, bias=True)
+            if cov.shape == ():  # JAX returns scalar for 1D input
+                cov = cov.reshape((1, 1))
+            inv_mass_matrix_sqrt = np.linalg.cholesky(cov)
+        else:
+            inv_mass_matrix_sqrt = np.std(zs, 0)
+        adapt_state = SAAdaptState(zs, pes, np.mean(zs, 0), inv_mass_matrix_sqrt)
+        k = categorical_logits(rng_key_z, np.zeros(zs.shape[0]))
+        z = unravel_fn(zs[k])
+        pe = pes[k]
+        sa_state = SAState(0, z, pe, 0., 0., False, adapt_state, rng_key_sa)
+        return device_put(sa_state)
+
+    def sample_kernel(sa_state, model_args=(), model_kwargs=None):
+        pe_fn = potential_fn
+        if potential_fn_gen:
+            pe_fn = potential_fn_gen(*model_args, **model_kwargs)
+        zs, pes, loc, inv_mass_matrix_sqrt = sa_state.adapt_state
+        rng_key, rng_key_z, rng_key_reject, rng_key_accept = random.split(sa_state.rng_key, 4)
+        z_old, unravel_fn = ravel_pytree(sa_state.z)
+
+        z = loc + _sample_proposal(inv_mass_matrix_sqrt, rng_key_z)
+        pe = pe_fn(unravel_fn(z))
+        pe = np.where(np.isnan(pe), np.inf, pe)
+        diverging = (pe - sa_state.potential_energy) > max_delta_energy
+
+        locs, scales = _get_proposal_loc_and_scale(zs, loc, inv_mass_matrix_sqrt, z)
+        zs_ = np.concatenate([z[None, :], zs])
+        pes_ = np.concatenate([pe[None], pes])
+        if inv_mass_matrix_sqrt.ndim == 2:  # dense_mass
+            log_weights = dist.MultivariateNormal(locs, scale_tril=scales).log_prob(zs_) + pes_
+        else:
+            log_weights = dist.Normal(locs, scales).log_prob(zs_).sum(-1) + pes_
+        log_weights = np.where(np.isnan(log_weights), -np.inf, log_weights)
+        j = categorical_logits(rng_key_reject, log_weights)
+        # XXX: numpy.delete is not available in JAX
+        zs = np.where(np.arange(zs.shape[0])[:, None] < j, zs_[:-1], zs_[1:])
+        pes = np.where(np.arange(pes.shape[0]) < j, pes_[:-1], pes_[1:])
+        loc = locs[j]
+        inv_mass_matrix_sqrt = scales[j]
+        adapt_state = SAAdaptState(zs, pes, loc, inv_mass_matrix_sqrt)
+
+        # XXX: we make a modification of SA sampler in [1]
+        # in [1], each MCMC state contains N points
+        # here we do resampling to pick randomly a point from N points
```

Comment:
can you please comment further?

Reply:
In Algorithm 1, line 10 of the paper, the author returns the collections of num_samples x N samples (denoted by z[k, n]). The posterior will contain all those points, and for each quantity `f`, `E(f(z)) = \int f(z)p(z)dz` will be computed as
`1/num_samples sum_k (1/N sum_n f(z[k,n]))`. I think that we can just simply compute that quantity by `1 / num_samples sum_k f(z[k])` where `z[k]` is randomly picked from N points `z[k,n]` (I think it is fine because they have uniform weights).

Reply:
why would you want to do this? you may lose information (samples)

Reply:
I mainly do this because of several reasons:
+ We want to keep the same MCMC API (otherwise, many logics will need to be revised so that the returned samples will have shape num_chains x num_samples x adapt_state_size x sample_shape). This is the main reason.
+ We don't need so many samples. Typically, SA will give us 100,000,000 - 1000,000,000  samples (from the experimental section of the reference) with num_samples=1000,000 and adapt_state_size=100-1000. I think 1000,000 samples are already large enough for prediction tasks. :D
+ I think if we want more samples, using multi-chains will be better. This SA is pretty suitable for the `vectorized` chain method.

Reply:
ok that makes sense. can we put a comment to that effect in the main docstring? something like "Note that unlike in reference [1], we return a randomly selected (i.e. thinned) subset of approximate posterior samples of size num_chains x num_samples instead of num_chains x num_samples x adapt_state_size."

Reply:
Thanks, I'll add that.